app/lib/shared/deps_graph.dart

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:logging/logging.dart';
+
+import 'package:pub_dartlang_org/frontend/models.dart';
+import 'package:pub_dartlang_org/shared/task_scheduler.dart';
+
+final Logger _logger = new Logger('pub.package_graph');
+
+// To avoid having the same package names many times in memory we canonicalize them.
+class StringCanonicalization {
+  final Map<String, String> _map = <String, String>{};
+
+  String canonicalize(String value) {
+    // Fast case is when [value] is already in the map.
+    final canonicalized = _map[value];
+    if (canonicalized != null) return canonicalized;
+
+    return _map[value] = value;
+  }
+}
+
+// TODO(kustermann): We could incorporate the pubspec constraints to make the
```

Comment:
I think this would be a lot of complexity here, for very little gain for now. I'd rather keep this class without version constraints, as long as the `analyzer` keeps up with the tasks.